srt-protocol/src/connection/mod.rs

```diff
@@ -188,55 +141,69 @@ impl DuplexConnection {
             Action::WaitForData(self.next_timer(now) - now)
         };
 
-        debug!(
-            "{:?}|{:?}|action - {:?}",
-            TimeSpan::from_interval(self.settings.socket_start_time, now),
-            self.settings.local_sockid,
-            action
-        );
-
+        self.debug("action", now, &action);
         action
     }
 
     pub fn is_open(&self) -> bool {
-        !(self.status.is_closed()
-            || (self.sender.is_closed() && self.receiver.is_flushed())
-            || self.receiver.is_closed())
+        self.status.is_open()
+    }
+
+    pub fn update_statistics(&mut self, now: Instant) {
+        self.statistics.elapsed_time = now - self.settings.socket_start_time;
     }
 
     pub fn next_packet(&mut self, now: Instant) -> Option<(Packet, SocketAddr)> {
-        let packet = self.sender.next_packet().or_else(|| {
-            self.receiver.next_packet().map(|p| {
-                self.sender.reset_keep_alive(now);
-                p
-            })
-        });
-        if let Some(p) = &packet {
-            debug!(
-                "{:?}|{:?}|send - {:?}",
-                TimeSpan::from_interval(self.settings.socket_start_time, now),
-                self.settings.local_sockid,
-                p
-            );
-        }
-        packet
+        self.output.pop_packet().map(|p| {
+            self.timers.reset_keepalive(now);
+            self.statistics.tx_all_packets += 1;
+            // payload length + (20 bytes IPv4 + 8 bytes UDP + 16 bytes SRT)
+            match &p {
+                Packet::Data(d) => {
+                    self.statistics.tx_data += 1;
+                    // payload length + (20 bytes IPv4 + 8 bytes UDP + 16 bytes SRT)
+                    self.statistics.tx_bytes += d.payload.len() as u64 + 44;
```

Comment:
Would prefer all these to be const globals

Reply:
I agree, yet I think we can do this later, when we better understand where it should go. I don't think we're done with this part of the code.